pkg/netscan/discover.go

```diff
@@ -0,0 +1,271 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package netscan
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"github.com/pkg/errors"
+	"math"
+	"net"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	NetworkUDP = "udp"
+	NetworkTCP = "tcp"
+)
+
+// AutoDiscover probes all addresses in the configured network to attempt to discover any possible
+// devices for a specific protocol
+func AutoDiscover(ctx context.Context, proto ProtocolSpecificDiscovery, params Params) []DiscoveredDevice {
+	params.Logger.Debugf("AutoDiscover called with the following parameters: %+v", params)
+	if len(params.Subnets) == 0 {
+		params.Logger.Warnf("Discover was called, but no subnet information has been configured!")
+		return nil
+	}
+
+	ipnets := make([]*net.IPNet, 0, len(params.Subnets))
+	var estimatedProbes int
+	for _, cidr := range params.Subnets {
+		if cidr == "" {
+			continue
+		}
+
+		ip, ipnet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			params.Logger.Errorf("Unable to parse CIDR %q: %s", cidr, err)
+			continue
+		}
+		if ip == nil || ipnet == nil || ip.To4() == nil {
+			params.Logger.Errorf("Currently only ipv4 subnets are supported. subnet=%q", cidr)
+			continue
+		}
+
+		ipnets = append(ipnets, ipnet)
+		// compute the estimate total amount of network probes we are going to make
+		// this is an estimate because it may be lower due to skipped addresses (existing devices)
+		sz, _ := ipnet.Mask.Size()
+		estimatedProbes += int(computeNetSz(sz))
+	}
+
+	if estimatedProbes == 0 {
+		params.Logger.Warnf("No valid CIDRs provided, unable to scan for devices.")
```

Comment:
```suggestion
		params.Logger.Warn("No valid CIDRs provided, unable to scan for devices.")
```

Reply:
That function doesn't exist on the interface. But if I switch to EdgeX logger I will remove the `f`.